### PR TITLE
CLC-5620 Create and populate Oec::CourseCode table and model

### PR DESCRIPTION
**Pull request by paulkerschen**
https://jira.ets.berkeley.edu/jira/browse/CLC-5651

After some dithering, I decided to use the opaque but machine-friendly five-letter L4 code to identify departments in the database. A `Berkeley::Departments` translation module follows the existing `Berkeley::Colleges` in using a simple in-memory map to deliver human-readable official department names without more mucking around in the DB.

`Berkeley::Colleges` gets a bit of cleanup on the side.

### CODE REVIEW

File: app/models/berkeley/colleges.rb [hunk 1]
```diff
@@ -2,38 +2,31 @@ module Berkeley
...
+        'ENGR' =>       'College of Engineering',
+        'ENV DSGN' =>   'College of Environmental Design',
+        'GRAD DIV' =>   'College of Letters & Science',
+        'JOURN' =>      'Graduate School of Journalism',
+        'L & S' =>      'College of Letters & Science',
+        'LAW' =>        'School of Law',
+        'NAT RES' =>    'College of Natural Resources',
+        'OPTMETRY' =>   'School of Optometry',
+        'PUB HLTH' =>   'School of Public Health',
+        'PUB POL' =>    'Richard & Rhonda Goldman School of Public Policy',
```
Comment (johncrossman) [inline app/models/berkeley/colleges.rb]: I believe it is 'Richard & Rhoda Goldman School of Public Policy'. Lord knows, we want neither Rhoda nor Rhonda contacting us.

Comment (paulkerschen) [inline app/models/berkeley/colleges.rb]: Thank you, @johncrossman, amended. I assume they regularly scour Github for their own names.

File: db/schema.rb [hunk 2]
```diff
@@ -153,6 +153,18 @@
...
+    t.string   "dept_name",      null: false
+    t.string   "catalog_id"
+    t.string   "dept_code",      null: false
+    t.boolean  "include_in_oec", null: false
+    t.datetime "created_at"
+    t.datetime "updated_at"
+  end
+
+  add_index "oec_course_codes", ["dept_code"], name: "index_oec_course_codes_on_dept_code", using: :btree
+  add_index "oec_course_codes", ["dept_name", "catalog_id"], name: "index_oec_course_codes_on_dept_name_and_catalog_id", unique: true, using: :btree
```
Comment (johncrossman) [inline db/schema.rb]: catalog_id is nullable and yet part of 'unique: true' index. Might that cause a problem?

File: spec/models/oec/course_code_spec.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
+describe Oec::CourseCode do
+
```
Comment (johncrossman) [inline spec/models/oec/course_code_spec.rb]: Perhaps a test where dept_name contains a single quote?

File: db/schema.rb (hunk 2 above)
Comment (paulkerschen) [inline db/schema.rb]: @johncrossman Excellent catch. Indeed, Postgres doesn't enforce unique constraints on null columns, and the intended constraint requires two partial indexes: 

http://stackoverflow.com/questions/8289100/create-unique-constraint-with-null-columns/8289253#8289253

I'll revise.

Comment (johncrossman) [inline db/schema.rb]: After a closer look I think the answer is "no problem".

File: spec/models/oec/course_code_spec.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    Oec::CourseCode.create(dept_name: 'BIOLOGY', catalog_id: '1AL', dept_code: 'IMMCB', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'BIOLOGY', catalog_id: '1B', dept_code: 'IBIBI', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'BIOLOGY', catalog_id: '1BL', dept_code: 'IBIBI', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'CATALAN', catalog_id: '1', dept_code: 'LPSPP', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'CHEM', catalog_id: '3A', dept_code: 'CCHEM', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'CHEM', catalog_id: '3B', dept_code: 'CCHEM', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'INTEGBI', catalog_id: '99B', dept_code: 'IBIBI', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'MCELLBI', catalog_id: '101', dept_code: 'IMMCB', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'PORTUG', catalog_id: '10', dept_code: 'LPSPP', include_in_oec: true)
+    Oec::CourseCode.create(dept_name: 'SPANISH', catalog_id: '200', dept_code: 'LPSPP', include_in_oec: false)
```
Comment (johncrossman) [inline spec/models/oec/course_code_spec.rb]: Why include_in_oec: false for SPANISH?

File: db/migrate/20150813225141_create_oec_course_codes.rb [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    add_index :oec_course_codes, :dept_code
+
+    reversible do |dir|
+      dir.up do
+        [
+          'A,RESEC', nil, 'MBARC', true,
+          'AEROSPC', nil, 'QLROT', false,
+          'AFRICAM', nil, 'SAAMS', false,
+          'AFRKANS', nil, 'HZGER', false,
+          'AGR CHM', nil, 'MEPMB', false,
```
Comment (johncrossman) [inline db/migrate/20150813225141_create_oec_course_codes.rb]: Should AGR CHM be 'true' since it's under MEPMB? It is a sibling of PLANTBI.

**Comment by johncrossman** [comment]
@paulkerschen  I'm done reviewing. Looks really good. These dept mappings make me say...
 
971002bb-acac-4c7f-bc80-61eb11954ec8

**Comment by paulkerschen** [comment]
@johncrossman Thanks for the considered review. I've pushed an amendment with changes:
- Partial indexes caused unnecessary problems switching between dev/prod Postgres and test sqlite, so wildcard `catalog_id` is now indicated by a blank string rather than `NULL`.
- `AGR CHEM` comes on board with its siblings.
- The `Oec::CourseCode` spec is fleshed out a bit more, including a department name with a real character, a hypothetical single course excluded from Spanish rather than the entire department, and examples illustrating the uniqueness constraints.

**Comment by johncrossman** [comment]
Thank you @paulkerschen. Elegant work on an inelegant set of requirements.

**Pull request merged**